### PR TITLE
ubkg api upgrade and refactoring

### PR DESCRIPTION
**Pull request by AlanSimmons**
Refactoring and upgrade of ubkg-api.

Release notes:https://github.com/x-atlas-consortia/ubkg-api/blob/simmons/neo4jv5/docs/v2%20Release%20Notes.md

- 18 new endpoints
- conversion of POST to GET

For all endpoints:
- extensive error handling and parameter validation
- throttling/guardrails to prevent unmanaged timeouts or OOME
- version compatibility checking

Use case: v5 instance of Data Distillery neo4j database.

### CODE REVIEW

File: ubkg-api-spec.yaml [hunk 1]
```diff
@@ -87,483 +100,1426 @@ paths:
...
+      summary: Returns basic information on the UBKG neo4j database.
+      operationId: database_info_server_get
+      responses:
+        '200':
+          description: basic information on the UBKG neo4j database
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/DatabaseInfo'
+  /node_types:
```
Comment (yuanzhou) [inline ubkg-api-spec.yaml:430]: Can we use hyphen `/node-types` as agreed with https://github.com/hubmapconsortium/devops/tree/main/rest-api-design#name-resource-with-lowercase-letters-and-hyphens-concisely

File: ubkg-api-spec.yaml [hunk 2]
```diff
@@ -87,483 +100,1426 @@ paths:
...
-                  $ref: '#/components/schemas/TermtypeCode'
-  /terms/{term_id}/concepts:
+                $ref: '#/components/schemas/SabCodeDetails'
+        '400':
+          description: invalid parameter name; non-numeric parameter value; negative parameter value
+        '404':
+          description: No Codes for the specified parameters
+        '5XX':
+          description: Unknown error
+  /sabs/term_types:
```
Comment (yuanzhou) [inline ubkg-api-spec.yaml:687]: Can we use hyphen for the fixed path?

File: ubkg-api-spec.yaml [hunk 3]
```diff
@@ -87,483 +100,1426 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SabTermTypes'
+        '400':
+          description: invalid parameter name; non-numeric parameter value; negative parameter value
+        '404':
+          description: No term types for the specified parameters
+        '5XX':
+          description: Unknown error
+  /semantics/semantic_types:
```
Comment (yuanzhou) [inline ubkg-api-spec.yaml:735]: Can we use hyphen for the fixed path?

File: ubkg-api-spec.yaml [hunk 4]
```diff
@@ -87,483 +100,1426 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SemanticTypes'
+        '400':
+          description: invalid parameter name; non-numeric parameter value; negative value
+        '404':
+          description: No subtypes of the specified Semantic Type.
+        '5XX':
+          description: Unknown error
+  /semantics/semantic_types/{identifier}:
```
Comment (yuanzhou) [inline ubkg-api-spec.yaml:767]: Can we use hyphen for the fixed path?

File: ubkg-api-spec.yaml [hunk 5]
```diff
@@ -87,483 +100,1426 @@ paths:
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SemanticTypes'
+        '400':
+          description: invalid parameter name; non-numeric parameter value; negative value
+        '404':
+          description: No subtypes of the specified Semantic Type.
+        '5XX':
+          description: Unknown error
+  /semantics/semantic_types/{identifier}/subtypes:
```
Comment (yuanzhou) [inline ubkg-api-spec.yaml:806]: Can we use hyphen for the fixed path?

File: ubkg-api-spec.yaml [hunk 6]
```diff
@@ -87,483 +100,1426 @@ paths:
...
-              rel: [isa, isa]
-              depth: 2
+                $ref: '#/components/schemas/ConceptPaths'
+        '400':
+          description: invalid parameter name; missing required parameter name; non-numeric parameter value; negative value; mindepth > maxdepth
+        '404':
+          description: No Concepts with paths originating from the specified concept and set of parameters. Includes responses from queries with execution times that exceed the maximum.
+        '5XX':
+          description: Unknown error
+  /concepts/{concept_id}/paths/trees:
```
[the remark this answers is not in the document]
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:257]: Corrected.

File: src/ubkg_api/instance/app.cfg [hunk 7]
```diff
@@ -0,0 +1,4 @@
+SERVER='bolt://localhost:4500'
+USERNAME='neo4j'
+PASSWORD='abcd1234'
+TIMEOUT=15000
```
[the remark this answers is not in the document]
Comment (AlanSimmons) [reply src/ubkg_api/instance/app.cfg]: I think that I fixed this. It took a bit of fiddling and playing hide-and-seek with the app.cfg file, but I no longer see it in the branch.

File: ubkg-api-spec.yaml (hunk 1 above)
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:430]: done

File: ubkg-api-spec.yaml [hunk 8]
```diff
@@ -87,483 +100,1426 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/NodeTypes'
+        '404':
+          description: No nodes types
+        '5XX':
+          description: Unknown error
+
+  /node_types/counts:
```
[the remark this answers is not in the document]
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:446]: done

File: ubkg-api-spec.yaml [hunk 9]
```diff
@@ -87,483 +100,1426 @@ paths:
...
-      operationId: concepts_concept_id_semantics_get
-      summary: Returns a list of semantic_types {Sty, Tui, Stn} of the concept
+      summary: This endpoint exists as a convenience to explain a limitation; a custom 400 message will explain the issue and suggest alternatives.
+      operationId: node_type_count_get
+      responses:
+        '400':
+          description: The response to this endpoint is likely to exceed timeout and so will not be attempted. Execute the node_types/(node_type) endpoint with the name of a node type (e.g., Codes). To obtain names of node types, execute the node_types endpoint.
+        '5XX':
+          description: Unknown error
+  /node_types/{node_type}/counts:
```
[the remark this answers is not in the document]
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:455]: done

File: ubkg-api-spec.yaml (hunk 2 above)
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:687]: done

File: ubkg-api-spec.yaml [hunk 10]
```diff
@@ -87,483 +100,1426 @@ paths:
...
-      operationId: terms_term_id_concepts_get
-      summary: Returns a list of concepts associated with the text string
+      operationId: sabs_term_types_get
+      summary: This endpoint exists as a convenience to explain a limitation; a custom 400 message will explain the issue and suggest alternatives.
+      responses:
+        '400':
+          description: The response to this endpoint cannot be run for all SABs because of memory limitations. Execute the /sabs/(sab)/term_types endpoint with the identifier for a single SAB. Execute the /sabs endpoint for a list of all SABs in the UBKG.
+        '5XX':
+          description: Unknown error
+  /sabs/{sab}/term_types:
```
[the remark this answers is not in the document]
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:696]: done

File: ubkg-api-spec.yaml (hunk 3 above)
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:735]: done

File: ubkg-api-spec.yaml (hunk 4 above)
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:767]: done

File: ubkg-api-spec.yaml (hunk 5 above)
Comment (AlanSimmons) [reply ubkg-api-spec.yaml:806]: done

**Review by yuanzhou — APPROVED** [review]
Will merge and do more testing on AWS once all the pieces are in place.

**Pull request merged**